src/ore/src/id_gen.rs

```diff
@@ -97,8 +108,155 @@ where
     }
 }
 
+/// A clone-able owned reference to an ID.
+///
+/// Once all of the [`IdHandle`]s referencing an ID have been dropped, we will then free the ID
+/// for later re-use.
+#[derive(Clone, Debug, PartialEq, Eq, Ord, PartialOrd)]
+pub enum IdHandle<T: From<u8> + AddAssign + PartialOrd + Copy> {
+    /// An ID "allocated" a compile time.
+    Static(T),
+    /// An ID allocated at runtime, get's freed once all handles have been dropped.
+    Dynamic(Arc<internal::IdHandleInner<T>>),
+}
+
+impl<T> IdHandle<T>
+where
+    T: From<u8> + AddAssign + PartialOrd + Copy,
+{
+    /// Creates a new [`IdHandle`] with the provided ID.
+    ///
+    /// Note: It is __entirely__ up to the caller to make sure the provided ID is not already being
+    /// used.
+    pub const fn new_static(id: T) -> Self {
+        IdHandle::Static(id)
+    }
+
+    /// Returns the ID associated with this [`IdHandle`].
+    pub fn val(&self) -> T {
```

Comment:
I like the idea, but since in this case `T` is `Copy`, it would be fairly trivial for a user to get a `T` anyways. The current places we use this function we need a `T`, so I don't think in this case returning a `&T` would get us much